### PR TITLE
Fixes #17465: Move backend data initialization to app initialization

### PR DESCRIPTION
**Pull request by ehelms**
Moves the data in backend systems that needs to be initialized,
Candlepin organizations and Pulp sync notifications, to application
initialization and out of the database seeds. This allows database
seeds to focus on the database solely, and to allo wthe application
to check on health services and whether data needs to be initialized
in a particular backend.

This has the effect of not requiring backends to be up during seed
process, but to be up during application start.

### CODE REVIEW

**Comment by ehelms** [comment]
I'm not super happy about the way I had to handle pings with the with the current design of the Ping class. Since this has an affect on the current Pipeline breakage, I went with what exists rather than doing a re-factoring prior to this change. I am willing to explore re-factoring if reviewers wish.

This tries to encapsulate the initialization at the later of the backend service and leave the init code in the engine definition itself as small as possible.

**Comment by stbenjam** [comment]
Looks reasonable to me, would be good to get @jlsherrill or someone familiar with this to take a look too

**Comment by jlsherrill** [comment]
Curious why you did it in server startup rather than a rake task in an
installer hook.

I'm a tad nervous about running the org creates on startup and also a bit
concerned that if candlepin takes too long to startup (which it has been
known to do) then the rails initialization would fail.

Has this been tested on a production install with using katello service
restart?

On Nov 23, 2016 1:25 PM, "Stephen Benjamin" <notifications@github.com>
wrote:

Looks reasonable to me, would be good to get @jlsherrill
<https://github.com/jlsherrill> or someone familiar with this to take a
look too

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/Katello/katello/pull/6470#issuecomment-262594346>, or mute
the thread

**Comment by ehelms** [comment]
I have not fully production tested this yet.

As for a rake task with installer hook, that seems like moving the functionality away from where it should live. This to me should be part of the applications responsibility - to ensure that backend data is in sync. If the worry is that the backends may not be up at application start time then I'd think we may want to look into a polling initialization that checks and waits for the backends to be accessible to run these tasks, triggered by application start.

Actions and potentially data views would be blocked by two conditions: the backend(s) are up and have been initialized.

**Comment by jlsherrill** [comment]
Another concern would be that the create task his a lock error.  We'd need
to handle that situation and not block the server from starting.

If it is going to live here, polling the ping would likely be a must, with
some sort of timeout, maybe five minutes?  I've seen candlepin take two to
three minutes, so five would give some overhead.   I think we'd need to
test this quite a bit in a production box and determine the user experience
when a backbends service is down.

I still think I prefer a rake task due to the risk and unknowns.  Blocking
the rails app from starting for minutes at a time seems a bit risky and
potentially difficult to diagnose for the user.  I see little upside
compared to a rake task and installer hook.

File: app/services/katello/candlepin/server.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+        return unless SETTINGS[:katello][:use_candlepin] && !Foreman.in_rake?('db:migrate')
+        candlepin_status = Katello::Ping.ping([:candlepin])
+
+        if candlepin_status[:status] == Katello::Ping::FAIL_RETURN_CODE
+          fail _("There was an issue with the Candlepin service %s: ") % candlepin_status[:candlepin][:message]
+        end
+
+        Organization.all.each do |org|
+          User.current = User.anonymous_admin
+          ForemanTasks.sync_task(::Actions::Katello::Organization::Create, org) unless org.library
```
Comment (ekohl) [inline app/services/katello/candlepin/server.rb]: Maybe you can check if there are orgs without a library before running the ping. That would save a ping on already initialized instances.

I also wonder: foreman tasks are async so they should be able to run at a later time, so why is the ping even needed? At that later time candlepin might even be unreachable for whatever reason.

File: app/services/katello/pulp/server.rb [hunk 2]
```diff
@@ -36,6 +36,17 @@ def self.config(url, user_remote_id)
...
+
+      def self.init
+        return unless SETTINGS[:katello][:use_pulp] && !Foreman.in_rake?('db:migrate')
+        pulp_status = Katello::Ping.ping(services: [:pulp])
+
+        if pulp_status[:status] == Katello::Ping::FAIL_RETURN_CODE
+          fail _("There was an issue with the Pulp service %s: ") % pulp_status[:services][:pulp][:message]
+        end
+
+        Katello::Repository.ensure_sync_notification
```
Comment (ekohl) [inline app/services/katello/pulp/server.rb]: Having a ```needs_sync_notification``` before pinging would increase reliability.

File: app/services/katello/candlepin/server.rb (hunk 1 above)
Comment (ehelms) [inline app/services/katello/candlepin/server.rb]: Foreman tasks can be async or sync, this one is doing it synchronously to ensure the creates happen before moving on. If we do this async, the user faces potential issues if the task hasn't finished but they are aloud to work on other entities such as an activation key.

Comment (ekohl) [inline app/services/katello/candlepin/server.rb]: Oh you're right and I could have seen that if I read it. That still leaves the option for a pre check avoiding a ping to an external service.

**Comment by ehelms** [comment]
What does it mean today when services are down for the user? There is not a lot they can do, and if they do try something they are hit with an error that they then have to fix. I see the downsides of the application not being able to come back up, but I also feel like I see an upside here of the user being presented with the backend services are down. And its quite obvious this needs to be fixed. 

An alternative could be to set a database setting similar to `db_seed_pending` to indicate if the backends have been initialized previously and skip if so. Then you are only worrying about it the first time the system is ever started. 

Another alternative is to move the Organization logic somewhere else, where it does a consistency check when accessed to detect if it needs to be synced to Candlepin. The Pulp one is harder cause we actually go to Pulp to check if the sync notification exists.

I'm worried a bit that a rake task + installer combo could lead to the web application being up and users thinking everything is OK if the rake task fails. Plus I was trying to avoid adding more steps that make Katello unique with respect to how Rails applications are initialized since this paradigm would be outside the usual realm of tasks run. 

I am with you on the concerns, I just like the idea that the web application ensures consistency. Perhaps the concerns are just more than is viable due to how these systems work unfortunately.

**Comment by ehelms** [comment]
Side question: does Foreman tasks have the functionality needed to move sync plans under our control yet from your knowledge? That alone would alleviate one of these but is probably a bigger change.

**Comment by ekohl** [comment]
There is no equivalent dependency within Foreman itself other than the database. All the other services it talks to are optional and that means there's no infrastructure to rely on/hook into here.

To me it sounds like the Candlepin part is the easier part. Would it make sense to tackle that first? Could it lazily initialize whenever Katello needs to interact with Candlepin?

**Comment by jlsherrill** [comment]
On Wed, Nov 23, 2016 at 2:28 PM, Eric D Helms <notifications@github.com>
wrote:

> What does it mean today when services are down for the user? There is not
> a lot they can do, and if they do try something they are hit with an error
> that they then have to fix. I see the downsides of the application not
> being able to come back up, but I also feel like I see an upside here of
> the user being presented with the backend services are down. And its quite
> obvious this needs to be fixed.
>
I agree, but i don't know that this PR fixes this.  This only ensures that
the services are up when rails is coming up, not that they are up
continuously.  Having a pulp worker die or candlepin die (due to qpid
issues) is not uncommon either.   I assume that apache would still be up
serving pulp content even if rails didn't start properly (probably want to
test to make 100% sure).


> An alternative could be to set a database setting similar to
> db_seed_pending to indicate if the backends have been initialized
> previously and skip if so. Then you are only worrying about it the first
> time the system is ever started.
>
I think that is much safer.  I would be more okay with this change if that
were the case.

**Comment by ehelms** [comment]
@ekohl we kinda do that today with our lazy_accessors for attributes that are needed to be read from the backend systems. For data manipulation, we do ping checks to ensure the system is up for Dynflow based actions. The situations we have here are just to clarify:

 1) All Organizations need a corresponding Owner object in Candlepin, this is generally only tricky for existing Organizations that would be present on an existing Foreman or with the current installation method, Foreman seeds an initial Organization and we need to ensure that gets created.

 2) Sync plans are handled solely by Pulp, so they kick off behind the scenes and we thus rely on Pulp to notify us when a sync has completed when performed via a sync plan. This is why we configure a notifier to let us know when one finishes to index the content and perform other actions we need for data consistency.

**Comment by ekohl** [comment]
> All Organizations need a corresponding Owner object in Candlepin, this is generally only tricky for existing Organizations that would be present on an existing Foreman or with the current installation method, Foreman seeds an initial Organization and we need to ensure that gets created.

It sounds like Candlepin runs fine if an org exists in Foreman without one in Candlepin. It's Foreman that expects it to exist. Would it be possible to make an organization without a corresponding owner object in Candlepin unusable or create it on the fly?

> Sync plans are handled solely by Pulp, so they kick off behind the scenes and we thus rely on Pulp to notify us when a sync has completed when performed via a sync plan. This is why we configure a notifier to let us know when one finishes to index the content and perform other actions we need for data consistency.

Maybe this could also be solved by making the feature unusable until it's configured. A clean Foreman / Katello has no sync plans configured so the notifier only needs to be in place when a user configures the first sync plan.

**Comment by ehelms** [comment]
@jlsherrill thoughts on @ekohl idea?

**Comment by jlsherrill** [comment]
:+1: to the idea of calling `Katello::Repository.ensure_sync_notification`  during sync plan creation.  I think that makes total sense.

I'm not sure about the second part I'd have a few questions.  

* What would happen when i tried to access a katello page on an org that wasn't in candlepin?
* Or registering a client?
* How would a user cause an org to be created in candlepin?
* Would this have to be done by the user even if katello was installed itself? (if its a manual process this could be confusing).

**Comment by ekohl** [comment]
> I'm not sure about the first part I'd have a few questions.

I thought about this not limited by knowledge of the current design and implementation. It makes perfect sense that the current implementation can't handle this and makes plenty of assumptions that it's there.

Is is possible to check using just the database if the candlepin org exists? If so, you can do pre-checks and not require a Candlepin connection on start up if the installation is configured. Requiring it on inconsistent (unmigrated) setups might be acceptable and a great step towards allowing a Foreman -> Katello migration.

You could even do it in an explicit rake task or keep it in for now. Currently pulp is causing the dependency cycle, not Candlepin.

**Comment by ehelms** [comment]
I've updated this to move the `ensure_sync_notification` to the sync plan `add_products` action and kept the current Candlepin interaction in the seeds file for now.

**Comment by jlsherrill** [comment]
ACK

**Pull request merged**